features/mbedtls/targets/TARGET_NUVOTON/TARGET_M480/ecp/ecp_internal_alt.c

```diff
@@ -0,0 +1,835 @@
+    /* Configure ECC point (X1, Y1) */
+    MBEDTLS_INTERNAL_MPI_NORM(&Np, P->X, N_, grp->P);
+    MBEDTLS_MPI_CHK(mbedtls_internal_mpi_write_eccreg(Np, (uint32_t *) CRPT->ECC_X1, NU_ECC_BIGNUM_MAXWORD));
+    MBEDTLS_INTERNAL_MPI_NORM(&Np, P->Y, N_, grp->P);
+    MBEDTLS_MPI_CHK(mbedtls_internal_mpi_write_eccreg(Np, (uint32_t *) CRPT->ECC_Y1, NU_ECC_BIGNUM_MAXWORD));
+
+    /* Configure ECC points (X2, Y2) */
+    if (eccop == ECCOP_POINT_ADD) {
+        MBEDTLS_INTERNAL_MPI_NORM(&Np, Q->X, N_, grp->P);
+        MBEDTLS_MPI_CHK(mbedtls_internal_mpi_write_eccreg(Np, (uint32_t *) CRPT->ECC_X2, NU_ECC_BIGNUM_MAXWORD));
+        MBEDTLS_INTERNAL_MPI_NORM(&Np, Q->Y, N_, grp->P);
+        MBEDTLS_MPI_CHK(mbedtls_internal_mpi_write_eccreg(Np, (uint32_t *) CRPT->ECC_Y2, NU_ECC_BIGNUM_MAXWORD));
+    }
+
+    crypto_ecc_prestart();
+    CRPT->ECC_CTL = (grp->pbits << CRPT_ECC_CTL_CURVEM_Pos) | eccop | CRPT_ECC_CTL_FSEL_Msk | CRPT_ECC_CTL_START_Msk;
+    ecc_done = crypto_ecc_wait();
+    
+    /* FIXME: Better error code for ECC accelerator error */
+    MBEDTLS_MPI_CHK(ecc_done ? 0 : -1);
+    
+    /* (X1, Y1) hold the normalized result. */
+    MBEDTLS_MPI_CHK(mbedtls_internal_mpi_read_eccreg(&R->X, (uint32_t *) CRPT->ECC_X1, NU_ECC_BIGNUM_MAXWORD));
+    MBEDTLS_MPI_CHK(mbedtls_internal_mpi_read_eccreg(&R->Y, (uint32_t *) CRPT->ECC_Y1, NU_ECC_BIGNUM_MAXWORD));
+    MBEDTLS_MPI_CHK(mbedtls_mpi_lset(&R->Z, 1));
+        
+cleanup:
+
+    mbedtls_mpi_free(&N_);
+    
+    return ret;
+}
+
+int mbedtls_internal_run_modop(mbedtls_mpi *r,
+                            const mbedtls_mpi *o1,
+                            const mbedtls_mpi *o2,
+                            const mbedtls_mpi *p,
+                            uint32_t pbits,
+                            uint32_t modop)
+{
+    if (r == NULL || 
+        o1 == NULL ||
+        o2 == NULL ||
+        p == NULL) {
+        return MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+    }
+    
+    /* Check o1/o2 are not negative */
+    if (mbedtls_mpi_cmp_int(o1, 0) < 0 ||
+        mbedtls_mpi_cmp_int(o2, 0) < 0) {
+        return MBEDTLS_ERR_MPI_NEGATIVE_VALUE;
+    }
+    
+    /* Check p is positive */
+    if (mbedtls_mpi_cmp_int(p, 0) <= 0) {
+        return MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+    }
+    
+    /* Check supported maximum key bits */
+    if (pbits > NU_ECC_MAXKEYBITS) {
+        return MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE;
+    }
+    
+    /* Check MODOP operations are legal */
+    if (modop != MODOP_DIV &&
+        modop != MODOP_MUL &&
+        modop != MODOP_ADD &&
+        modop != MODOP_SUB) {
+        return MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+    }
+    
+    int ret;
+    bool ecc_done;
+
+    mbedtls_mpi N_;
+    const mbedtls_mpi *Np;
+    
+    mbedtls_mpi_init(&N_);
+    
+    /* Use MBEDTLS_INTERNAL_MPI_NORM(Np, N1, N_, P) to get normalized MPI
+     *
+     * N_: Holds normalized MPI if the passed-in MPI N1 is not
+     * Np: Pointer to normalized MPI, which could be N1 or N_
+     */
```

Comment:
I don't see the benefit of this macro compared to performing a modular reduction on the respective variable in-place. On the functional side it's the same, on the performance side it has the chance to be less costly because it is not enforced that the bignum module will need to make a copy, and it would be more readable (for my taste). Could you elaborate why you chose this path, or change `MBEDTLS_INTERNAL_MPI_NORM(a,b,c,d)` to `MBEDTLS_INTERNAL_MPI_NORM(x,N)` performing a modular reduction on `x` mod `N` in-place if necessary? 

Looking at it, I think you can even reduce this to a call to `mbedtls_mpi_mod_mpi( x, x, N )` everywhere, as the latter function checks if already `x < N` before performing any expensive operations.

Reply:
Oh, I see, many MPI's are unfortunately declared `const` in the fixed API, and we are not allowed to perform modular reduction in-place. That makes the necessity and motivation for the macro clearer.